### PR TITLE
WIP switch to fetch by default

### PR DESCRIPTION
**Pull request by daleharvey**
Just posting as a WIP to get comments / help etc

This is passing all basic tests except https://github.com/pouchdb/pouchdb/blob/master/tests/integration/test.basics.js#L131, fairly confused at that failure as I cant see anywhere we would add the statuscode to a successful http request (the bulkdocs request is successful, the error is called because the first and only result is an error)

### CODE REVIEW

**Comment by daleharvey** [comment]
Fixed above error, just need to go through one by one and fix the edge case errors now

**Comment by daleharvey** [comment]
Getting close, tests are 100% passing now, the skipped tests are either missing api's I need to readd, or use the `_ajax` trick that needs rewritten (migration test failures are confusing)

**Comment by daleharvey** [comment]
actually migration tests failing make sense, was testing against couch master, this is all working in the browser (just need to fix travis versions) and fix for nodejs

**Comment by marten-de-vries** [comment]
I'm actually currently trying to get pouchdb-server's unit tests to run with a pouchdb from this branch to get a better view of the consequences. I'm currently using  [fetch-cookie](https://www.npmjs.com/package/fetch-cookie). Can I assume you have something similar planned?

**Comment by marten-de-vries** [comment]
Oh, if you do, ``fetch-cookie`` seems broken when using ``Headers``. I'm now using the underlying ``node-fetch``, but that means no cookie support. That might actually be fine from a PouchDB perspective, and a problem to leave to plugins that need it.

**Comment by daleharvey** [comment]
I didnt plan for node yet, fetch-cookie isnt looking super active / stable, I would prefer to have cookie support built in, a fetch wrapper around request would be pretty simple, kinda surprised there isnt a mature implementation for fetch as yet

**Comment by marten-de-vries** [comment]
[node-fetch](https://www.npmjs.com/package/node-fetch) itself is well-maintained and even pretty complete, they just decided [a cookie jar is out of scope for the project](https://github.com/bitinn/node-fetch/issues/94). Writing a ``fetch`` implementation on top of ``request`` supporting what PouchDB needs only should work. One implementing the complete spec would probably be a project on its own.

**Comment by daleharvey** [comment]
Yeh I meant a mature implementation that included cookie support, I think it should be trivial / out of the box to have cookies working so I am thinking a request wrapper is probably fine, its a pretty small wrapper as we dont hit a lot of the API (we currently already do the opposite, map request calls to fetch / or xhr). Its also not a super critical which implementation we use since it can now easily be swapped out

**Comment by daleharvey** [comment]
CI tests are passing in firefox now :) ready to start working on node, I think I will do a request wrapper unless anyone thinks there is a better way

**Comment by daleharvey** [comment]
The request wrapper was more complicated than I expected, using node-fetch for now :)

**Comment by daleharvey** [comment]
Mostly working in node, if anyone wanted to debug the attachment failures I hate those bugs :P

**Comment by daleharvey** [comment]
Fixed the attachment bugs, now working on the override again

``` 
new PouchDB('name', { 
  fetch: function (url, opts) { 
    return fetch(url, opts);
  }
});
```
works in the browser, but obviously not in node, I am thinking we can either do `fetch: function (url, opts, fetch)` or `fetch: function (url, opts) { return PouchDB.fetch .. `, leaning towards the latter

**Comment by daleharvey** [comment]
ok, trying to make a `pouchdb-fetch` module however running into troubles with the build system, error is https://gist.github.com/daleharvey/c321082a4a7b7645323a3240190fa6ad, @nolanlawson or anyone else familiar with the build system know what I need to do here?

Also Nolan do you have a script we can check in to make sure people have access to all pouchdb-repos at publish time? cheers

**Comment by marten-de-vries** [comment]
Sorry, no idea where ``PouchdB.ajax`` was set, and I don't recognize that error either. Nice progress, though!

**Comment by daleharvey** [comment]
rollup was being super weird about variable names, fixed now and passing in node and firefox :) getting close

**Comment by daleharvey** [comment]
Hrm, The map reduce tests are failing in the browser due to attatchments, not surprising but I cant reproduce

**Comment by daleharvey** [comment]
ok so the major things left is ie is failing unexpected and the fetch cookie test is failing, I am on holiday till the 17th and if anyone wanted to pick those up it would be amazing :) otherwise will get to this when I get back but its super close at the moment

**Comment by ptitjes** [comment]
@daleharvey An easy way is to use `fetch-cookie`. I know you said it seems not well-maintained but it seems the most maintained from all the alternatives. Also, regarding @marten-de-vries's comment, the redirect problem seems to be handled correctly on Node.js by `fetch-cookie` [thanks to the `node-fetch` API](https://github.com/valeriangalliat/fetch-cookie#cookies-and-redirection).

Also, as a bonus, it uses `tough-cookie` so we could easily ship a mechanism to customize the cookie jar inside `pouchdb-fetch`.

So I tried it and can confirm that the fetch cookie test is passing by using `fetch-cookie`. Do you want that I create a branch above your `fetch` branch with this commit ?

**Comment by marten-de-vries** [comment]
@ptitjes I actually got a crash when I tried it. Something with the ``Headers`` object, but if you managed to get it to work, apparently I did something stupid or you or @daleharvey fixed it somehow.

If you could open a PR with your code (with 'WIP' in the title to prevent accidental merging), it could go through a full Travis run. In a separate branch is fine, it can be ``git cherry-pick``ed into this branch if it passes. Personally, I think it would be an acceptable solution if Travis passes, as ``fetch-cookie`` is mostly a thin wrapper. But @daleharvey is probably in a better position to make a final decision on if it's maintained well enough.

**Comment by daleharvey** [comment]
Yeh node-fetch doesnt look well maintained, but we can send PR's and if they dont get fixed we can maintain a fork, tried giving it a shot but not sure how its supposed to work, @marten-de-vries how dod you test it? 

I tried 

> import fetch as nodeFetch from 'node-fetch';
> import fetchCookie from 'fetch-cookie';
> var fetch = fetchCookie(nodeFetch);

but gives me a build error

**Comment by marten-de-vries** [comment]
I never got ``fetch-cookie`` to work, only ``node-fetch``, so no idea. @ptitjes could you tell us what you did/push the code somewhere?

**Comment by ptitjes** [comment]
@daleharvey @marten-de-vries Sorry guys I had busy time and only see your comments now.

I can push a PR during the week-end. In the mean time, here is the diff:
```diff
Index: packages/node_modules/pouchdb-fetch/src/fetch.js
===================================================================
--- packages/node_modules/pouchdb-fetch/src/fetch.js	(date 1512220316000)
+++ packages/node_modules/pouchdb-fetch/src/fetch.js	(date 1518509336000)
@@ -2,6 +2,9 @@
 
 import fetch from 'node-fetch';
 
+// Use fetch-cookie for now
+var fetchWithCookies = require('fetch-cookie/node-fetch')(fetch);
+
 /* We can fake the abort, the http adapter keeps track
    of ignoring the result */
 function AbortController() {
@@ -9,4 +12,4 @@
 }
 
 var Headers = fetch.Headers;
-export {fetch, Headers, AbortController};
+export {fetchWithCookies as fetch, Headers, AbortController};
Index: package.json
===================================================================
--- package.json	(date 1512220316000)
+++ package.json	(date 1518508855000)
@@ -40,6 +40,7 @@
     "clone-buffer": "1.0.0",
     "debug": "3.1.0",
     "double-ended-queue": "2.1.0-0",
+    "fetch-cookie": "^0.7.0",
     "fruitdown": "1.0.2",
     "immediate": "3.0.6",
     "inherits": "2.0.3",
```

With that I have the component tests passing, though I did not try the full suite.

**Comment by ptitjes** [comment]
OK, I see that you figured it out...
